File: incre-tests/synduce/misc/count_between.f
Inductive Tree = leaf Int | node {Int, Tree, Tree};

max = \a: Int. \b: Int. if < a b then b else a;
min = \a: Int. \b: Int. if < a b then a else b;

tmin = fix (
  \f: Tree -> Int. \t: Tree.
  match t with
    leaf w -> w
  | node {w, l, r} -> min w (min (f l) (f r))
  end
);

tmax = fix (
  \f: Tree -> Int. \t: Tree.
  match t with
    leaf w -> w
  | node {w, l, r} -> max w (max (f l) (f r))
  end
);

@Input lo: Int;
@Input hi: Int;

repr = fix (
  \f: Tree -> Compress Tree. \t: Tree.
  match t with
    leaf w -> leaf w
  | node {w, l, r} ->
    if (>= w hi) then node {w, f l, f r} else node {w, f l, f r}
  end
);

spec = fix (
  \f: Tree -> Bool. \t: Tree.
  match t with
    leaf w -> (and (< lo w) (< w hi))
  | node {w, l, r} -> or (or (and (< lo w) (< w hi)) (f l)) (f r)
  end
);

main = \t: Tree. if < lo hi then spec (repr t) else false;

File: ml-benchmark/synduce/misc/count_between.ml
type 'a tree =
  | Leaf of 'a
  | Node of 'a * 'a tree * 'a tree

let min a b = if a < b then a else b
let max a b = if a > b then a else b

let rec tree_min = function
  | Leaf x -> x
  | Node (a, l, r) -> min a (min (tree_min l) (tree_min r))

let rec tree_max = function
  | Leaf x -> x
  | Node (a, l, r) -> max a (max (tree_max l) (tree_max r))

@Input val lo: int 
@Input val hi: int

let rec spec = function 
| Leaf a -> lo < a && a < hi
| Node (a, l, r) -> (lo < a && a < hi) || spec l || spec r

val repr: int tree -> (int tree) compress
let rec repr = function
| Leaf a -> Leaf a 
| Node (a, l, r) -> Node (a, repr l, repr r)

let prog t = spec (repr t)
